ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml

```diff
@@ -237,57 +237,68 @@ let dss_vcpus xc doms =
-             )
-          :: dss
-        with _ -> dss
+        let dom_cpu_time =
+            Int64.(
+              to_float @@ logand dom.Xenctrl.cpu_time xen_flag_complement
+            )
+          in
+          let dom_cpu_time = dom_cpu_time /. 1.0e9 in
+          try
+            let ri = Xenctrl.domain_get_runstate_info xc domid in
+            ( Rrd.VM uuid
+            , Ds.ds_make ~name:"runstate_fullrun" ~units:"(fraction)"
+                ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time0 /. 1.0e9))
+                ~description:"Fraction of time that all VCPUs are running"
+                ~ty:Rrd.Derive ~default:false ~min:0.0 ()
+            )
+            :: ( Rrd.VM uuid
+              , Ds.ds_make ~name:"runstate_full_contention" ~units:"(fraction)"
+                  ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time1 /. 1.0e9))
+                  ~description:
+                    "Fraction of time that all VCPUs are runnable (i.e., \
+                      waiting for CPU)"
+                  ~ty:Rrd.Derive ~default:false ~min:0.0 ()
+              )
+            :: ( Rrd.VM uuid
+              , Ds.ds_make ~name:"runstate_concurrency_hazard"
+                  ~units:"(fraction)"
+                  ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time2 /. 1.0e9))
+                  ~description:
+                    "Fraction of time that some VCPUs are running and some are \
+                      runnable"
+                  ~ty:Rrd.Derive ~default:false ~min:0.0 ()
+              )
+            :: ( Rrd.VM uuid
+              , Ds.ds_make ~name:"runstate_blocked" ~units:"(fraction)"
+                  ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time3 /. 1.0e9))
+                  ~description:
+                    "Fraction of time that all VCPUs are blocked or offline"
+                  ~ty:Rrd.Derive ~default:false ~min:0.0 ()
+              )
+            :: ( Rrd.VM uuid
+              , Ds.ds_make ~name:"runstate_partial_run" ~units:"(fraction)"
+                  ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time4 /. 1.0e9))
+                  ~description:
+                    "Fraction of time that some VCPUs are running, and some are \
+                      blocked"
+                  ~ty:Rrd.Derive ~default:false ~min:0.0 ()
+              )
+            :: ( Rrd.VM uuid
+              , Ds.ds_make ~name:"runstate_partial_contention"
+                  ~units:"(fraction)"
+                  ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time5 /. 1.0e9))
+                  ~description:
+                    "Fraction of time that some VCPUs are runnable and some are \
+                      blocked"
+                  ~ty:Rrd.Derive ~default:false ~min:0.0 ()
+              )
+            :: ( Rrd.VM uuid
+              , Ds.ds_make ~name:(Printf.sprintf "cpu_usage") ~units:"(fraction)"
+                  ~description:(Printf.sprintf "domain: %d CPU usage" domid)
```

Comment:
The rest of the metrics avoid mentioning domains, probably because it leaks abstraction. For example, whenever a VM migrates it will the description will change, this can affect users consolidating stats

Reply:
That make sense!  I will delete the domain id from description.